lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs

```diff
@@ -0,0 +1,1015 @@
+  where
+    balanceAvailable = view #balance index
+    balanceRequested = adjustAllQuantities (* 2) balanceAvailable
+
+prop_runSelection_UTxO_exactlyEnough
+    :: Small UTxOIndex
+    -> Property
+prop_runSelection_UTxO_exactlyEnough (Small index) = monadicIO $ do
+    SelectionState {selected, leftover} <-
+        run $ runSelection Nothing index balanceRequested
+    let balanceSelected = view #balance selected
+    let balanceLeftover = view #balance leftover
+    assert $ balanceSelected == balanceRequested
+    assert $ balanceLeftover == TokenBundle.empty
+  where
+    balanceRequested = view #balance index
+
+prop_runSelection_UTxO_extraSourceUsed
+    :: Maybe Coin
+    -> Small UTxOIndex
+    -> Property
+prop_runSelection_UTxO_extraSourceUsed extraSource (Small index) =
+    monadicIO $ case almostEverything of
+        Nothing ->
+            assert True
+        Just balanceRequested | balanceRequested == TokenBundle.empty -> do
+            SelectionState {selected,leftover} <-
+                run $ runSelection extraSource index balanceRequested
+            let balanceSelected = view #balance selected
+            let balanceLeftover = view #balance leftover
+            assert $ balanceLeftover == view #balance index
+            assert $ balanceSelected == TokenBundle.empty
+        Just balanceRequested -> do
+            monitor (cover 80 True "sometimes there are Ada")
+            SelectionState {selected} <-
+                run $ runSelection extraSource index balanceRequested
+            let balanceSelected = view #balance selected
+            let coinSelected = TokenBundle.coin $
+                    addExtraSource extraSource balanceSelected
+            monitor $ counterexample $ unlines
+                [ "balance selected: " <> show balanceSelected
+                ]
+            assert $ coinSelected >= TokenBundle.coin balanceRequested
+            assert $ balanceSelected /= TokenBundle.empty
+  where
+    almostEverything = TokenBundle.subtract
+        (view #balance index)
+        (TokenBundle.fromCoin (Coin 1))
+
+prop_runSelection_UTxO_moreThanEnough
+    :: Maybe Coin
+    -> Small UTxOIndex
+    -> Property
+prop_runSelection_UTxO_moreThanEnough extraSource (Small index) = monadicIO $ do
+    SelectionState {selected, leftover} <-
+        run $ runSelection extraSource index balanceRequested
+    let balanceSelected = view #balance selected
+    let balanceLeftover = view #balance leftover
+    monitor $ cover 80
+        (assetsRequested `Set.isProperSubsetOf` assetsAvailable)
+        "assetsRequested ⊂ assetsAvailable"
+    monitor $ cover 50 (Set.size assetsRequested >= 4)
+        "size assetsRequested >= 4"
+    monitor $ counterexample $ unlines
+        [ "balance available: " <> show balanceAvailable
+        , "balance requested: " <> show balanceRequested
+        , "balance selected:  " <> show balanceSelected
+        , "balance leftover:  " <> show balanceLeftover
+        ]
```

Comment:
Here we can take advantage of the `Buildable` instance for `TokenBundle`, which produces output that's easier to read than `Show`:

```suggestion
    monitor $ counterexample $ unlines
        [ "balance available:"
        , pretty (TokenBundle.Nested balanceAvailable)
        , "balance requested:"
        , pretty (TokenBundle.Nested balanceRequested)
        , "balance selected:"
        , pretty (TokenBundle.Nested balanceSelected)
        , "balance leftover:"
        , pretty (TokenBundle.Nested balanceLeftover)
        ]
```

The `Nested` instance produces output that's a bit more concise, as assets are grouped by token policy.

Reply:
Indeed! Good point.